Dev Debugging docs (#5480)
opencast/opencast: docs/guides/developer/docs/develop/debugging.md

# Debugging in Opencast

This guide will show you how to debug Opencast.

## Debugging strategies

There are different ways to debug Opencast. You can debug the backend and the frontend.

> The backend is the Java code while the frontend is the Angular code.

Reviewer: Angular is departing the codebase, hopefully soon.  The new one is TypeScript, IIRC.

### How to know what to debug?

> First, you want to know you to identify where is the problem. The frontend problems are mostly visible in the browser.

Reviewer: `First, you need to determine where the problem is.`

The frontend and the backend are connected. The frontend calls the backend via REST endpoints.
If the information from the API is not correct, is a problem of the backend.
If the information is correct but, the frontend is not showing the information correctly, is a problem of the frontend.

## Backend debugging

### Logging configuration

The logging configuration is in the `etc/org.ops4j.pax.logging.cfg` file.
You can change the logging level of a package by adding the following line:

    log4j.logger.org.opencastproject.capture=DEBUG

Where `org.opencastproject.capture` is the package name and `DEBUG` is the logging level.

The logging level can be one of the following:

* `TRACE`
* `DEBUG`
* `ERROR`
* `WARN`
* `INFO`

### Attaching a Remote Debugger

There are different ways to debug the backend of Opencast.
The easiest way is to attach a remote debugger here are some two ways to do that.

#### For source installations

To debug a running Opencast system, you can attach a remote debugger in your IDE (Eclipse, NetBeans, etc.). For that
you have to enable the remote debugging in Karaf OSGI server that runs Opencast.

> You have to add “debug” as an additional parameter to the Opencast start script:

Reviewer: Smart quotes on debug should be reverted.

Author: Done, LanguageTool (The European Grammarly) always suggests me to change to smart quotes, is there any difference for Markdown?

    export KARAF_DEBUG=true

> In case you need to change the port for debugging, you can add this another line:

Reviewer: ```suggestion
In case you need to change the port for debugging, you can add another line:
```

For more information on remote debugging with Karaf you can visit [this
site.](https://karaf.apache.org/manual/latest/#_debugging)

> It is **not recommended** enabling remote debugging on production systems!

Reviewer: ```suggestion
Enabling remote debugging on production systems is **not recommended**!
```

or

```suggestion
Remote debugging is **not recommended** on production systems!
```

### Adding Dummy Capture Agent

> In the case that is needed, you can create a dummy capture agent with testing porpoises.

Reviewer: ```suggestion
In the case that is needed, you can create a dummy capture agent with testing purposes.
```

Click on `submit` and is ready to go.

> Additional you can call this REST endpoint directly using `CURL` for example.

Reviewer: curl, like ffmpeg, is all lower case.

```suggestion
Additional you can call this REST endpoint directly using `curl` for example.
```

## Frontend debugging

> To debug the frontend you can use the browser developer tools. For example in Chrome you can open the developer tools.

Reviewer: ```suggestion
To debug the frontend you can use the browser developer tools. For most browsers this is bound to your F12 key by default.
```
